x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/GrokPatternCreator.java

```diff
@@ -0,0 +1,600 @@
+     *                    can be appended by this method.
+     * @param sampleMessages Sample messages that any non-<code>null</code> return will match.
+     * @param mappings Will be updated with mappings appropriate for the returned pattern, if non-<code>null</code>.
+     * @return A tuple of (time field name, Grok string), or <code>null</code> if no suitable Grok pattern was found.
+     */
+    public static Tuple<String, String> findFullLineGrokPattern(List<String> explanation, Collection<String> sampleMessages,
+                                                                Map<String, Object> mappings) {
+
+        for (FullMatchGrokPatternCandidate candidate : FULL_MATCH_GROK_PATTERNS) {
+            if (candidate.matchesAll(sampleMessages)) {
+                return candidate.processMatch(explanation, sampleMessages, mappings);
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Build a Grok pattern that will match all of the sample messages in their entirety.
+     * @param explanation List of reasons for making decisions.  May contain items when passed and new reasons
+     *                    can be appended by this method.
+     * @param sampleMessages Sample messages that the returned Grok pattern will match.
+     * @param seedPatternName A pattern that has already been determined to match some portion of every sample message.
+     * @param seedFieldName The field name to be used for the portion of every sample message that the seed pattern matches.
+     * @param mappings Will be updated with mappings appropriate for the returned pattern, excluding the seed field name.
+     * @return The built Grok pattern.
+     */
+    public static String createGrokPatternFromExamples(List<String> explanation, Collection<String> sampleMessages, String seedPatternName,
+                                                       String seedFieldName, Map<String, Object> mappings) {
+
+        GrokPatternCandidate seedCandidate = new NoMappingGrokPatternCandidate(seedPatternName, seedFieldName);
+
+        Map<String, Integer> fieldNameCountStore = new HashMap<>();
+        StringBuilder overallGrokPatternBuilder = new StringBuilder();
+
+        processCandidateAndSplit(explanation, fieldNameCountStore, overallGrokPatternBuilder, seedCandidate, true, sampleMessages, mappings,
+            false, 0, false, 0);
+
+        return overallGrokPatternBuilder.toString().replace("\t", "\\t").replace("\n", "\\n");
+    }
+
+    /**
+     * Given a chosen Grok pattern and a collection of message snippets, split the snippets into the
+     * matched section and the pieces before and after it.  Recurse to find more matches in the pieces
+     * before and after and update the supplied string builder.
+     */
+    private static void processCandidateAndSplit(List<String> explanation, Map<String, Integer> fieldNameCountStore,
```

Comment:
The long arg lists in these methods make me wonder if we could refactor as an object whose members capture those parameters. Hard to come with a more concrete suggestion without trying to do it, but these long arg lists and static methods set make me wonder if we could simplify by going more OO here. This class seems like it is doing a lot and could be split up in several smaller classes with reduced responsibilities.

Reply:
Good idea - it's done in https://github.com/elastic/elasticsearch/pull/32788/commits/37a92f7a67f762fbfbbc92d53db807761b5b0ef2

This class got progressively more complex over time.  When I started out it made sense to just have a couple of static methods but now you're right it needs to be an object with members.